src/sql/src/session/vars2.rs

```diff
@@ -0,0 +1,685 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::fmt::{self, Debug};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use mz_build_info::BuildInfo;
+use mz_dyncfg::{ConfigSet, ConfigType, ConfigVal};
+use once_cell::sync::Lazy;
+use uncased::UncasedStr;
+
+use crate::session::user::{User, SUPPORT_USER, SYSTEM_USER};
+use crate::session::vars::{all_dyn_configs, ConnectionCounter, EndTransactionAction, Var};
+use crate::session::vars2::constraints::{DomainConstraint, ValueConstraint};
+use crate::session::vars2::definitions::FeatureFlag;
+use crate::session::vars2::errors::{VarError, VarErrorKind};
+
+mod constraints;
+mod definitions;
+mod errors;
+mod polyfill;
+mod value;
+
+use definitions::*;
+use polyfill::LazyValueFn;
+use value::Value;
+
+/// Definition of a Session or System variable.
+#[derive(Clone)]
+pub struct VarDefinition {
+    /// Name of the variable, case-insensitive matching.
+    name: &'static UncasedStr,
+    /// Description of the variable.
+    description: &'static str,
+    /// TODO(parkmcar): What does internal mean?
+    internal: bool,
+
+    /// Default compiled in value for this variable.
+    value: VarDefaultValue,
+    /// Constraint that must be upheld for this variable to be valid.
+    constraint: Option<ValueConstraint>,
+    /// Optionally hides this variable if it's related to a feature flag being enabled.
+    feature_flag: Option<&'static FeatureFlag>,
+
+    /// Method to parse [`VarInput`] into a type that implements [`Value`].
+    ///
+    /// The reason `parse` exists as a function pointer is because we want to achieve two things:
+    ///   1. `VarDefinition` has no generic parameters.
+    ///   2. `Value::parse` returns an instance of `Self`.
+    /// `VarDefinition` holds a `dyn Value`, but `Value::parse` is not object safe because it
+    /// returns `Self`, so we can't call that method. We could change `Value::parse` to return a
+    /// `Box<dyn Value>` making it object safe, but that creates a footgun where it's possible for
+    /// `Value::parse` to return a type that isn't `Self`, e.g. `<String as Value>::parse` could
+    /// return a `usize`!
+    ///
+    /// So to prevent making `VarDefinition` generic over some type `V: Value`, but also defining
+    /// `Value::parse` as returning `Self`, we store a static function pointer to the `parse`
+    /// implementation of our default value.
+    parse: fn(VarInput) -> Result<Box<dyn Value>, VarErrorKind>,
```

Comment:
One thing that I don't really understand is, what's to stop `value` from containing something like `VarDefaultValue::Static(&'static dyn <String as Value>)` (I have no idea if that syntax is valid but hopefully the idea is clear) but we initialize `parse` to some function that returns a `usize`? Are we just relying on whoever initializes this struct to provide a matching `parse` function?

Reply:
> One thing that I don't really understand is, what's to stop value from containing something like `VarDefaultValue::Static(&'static dyn <String as Value>)`

Unfortunately you can't add that type info, `dyn <String as Value>` without making the struct generic over some type `V`, which when then fall into the same issue as before. 

> Are we just relying on whoever initializes this struct to provide a matching parse function?

A TODO I have is to cleanup the visibilities on `VarDefinition` and move it into a "sealed" module, but given some fields on `VarDefinition` are private you can't create one directly, and need to use the `new_*(...)` methods which enforce the `parse` function is correct